src/py_backed.rs

```diff
@@ -0,0 +1,117 @@
+use std::ops::Deref;
+
+use crate::{
+    types::{
+        any::PyAnyMethods, bytearray::PyByteArrayMethods, bytes::PyBytesMethods,
+        string::PyStringMethods, PyByteArray, PyBytes, PyString,
+    },
+    Bound, DowncastError, FromPyObject, Py, PyAny, PyResult,
+};
+
+/// A wrapper around `str` where the storage is owned by a Python `bytes` or `str` object.
+///
+/// This type gives access to the underlying data via a `Deref` implementation.
+pub struct PyBackedStr {
+    #[allow(dead_code)]
+    storage: PyBackedStrStorage,
```

Comment:
Is there any upside here to the enum instead of just storing `Py<PyAny>`? The only thing mattering here after construction is decrementing the reference count, isn't it?

Reply:
Yep, originally I was thinking there might be value in the better type information, but for now let's just simplify 👍